Handle negative surface observation uncertainty values

supriyamantri opened the pull request:
* **Handle negative error values for flask data correctly using NaN**

This PR updates `get_obs_surface` so negative repeatability/variability values used as missing-data flags are converted to `NaN` before optional averaging. Uncertainty variables that become entirely `NaN` after that conversion are dropped, while unrelated all-`NaN` variables are left untouched.

* **Please check if the PR fulfills these requirements**

- [x] Closes #1538
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing)
- [x] Changelog entry added

*Note: if any of the above are not needed for a PR please separate to below and remove the checkbox.*

supriyamantri committed:
WIP: negative error handling fix in _access

supriyamantri committed:
Format retrieve/_access.py with black

brendan-m-murphy commented on openghg/retrieve/_access.py:
```diff
@@ -232,6 +232,28 @@ def get_obs_surface(
+            logger.info(
+                f"{var_to_delete} contain only nan for obs. in {surface_keywords}. They are thus deleted."
+            )
+            data = data.drop_vars(var_to_delete)
```
This is a good idea, and will probably help with the potential issues of using a dask backed array in `.where` (which doesn't happen in the other branch because `.compute` is called).

brendan-m-murphy commented:
@copilot apply changes based on the comments in [this thread](https://github.com/openghg/openghg/pull/1544#pullrequestreview-4119860243)

Copilot committed:
Refactor: extract _sanitise_negative_uncertainties helper, fix dask bool, fix NaN casing, add tests

Agent-Logs-Url: https://github.com/openghg/openghg/sessions/42f10fa7-39ca-48e6-8215-50b910d02985

Co-authored-by: brendan-m-murphy <11130776+brendan-m-murphy@users.noreply.github.com>

Copilot committed:
Apply code review suggestions: rename vars_to_delete and surface_keywords

Agent-Logs-Url: https://github.com/openghg/openghg/sessions/42f10fa7-39ca-48e6-8215-50b910d02985

Co-authored-by: brendan-m-murphy <11130776+brendan-m-murphy@users.noreply.github.com>

brendan-m-murphy commented on openghg/retrieve/_access.py:
```diff
@@ -88,6 +88,38 @@ def _get_generic(
+        dv for dv in data.data_vars if str(dv).endswith("repeatability") or str(dv).endswith("variability")
+    ]
+    for dv in uncertainty_data_vars:
+        data[dv] = data[dv].where(data[dv] >= 0.0)
```
@copilot  this might raise an error if data[dv] is a lazy-loaded dask array. This is taken care of in the `average is not None` branch because the data is computed there. The only certain way around this is to compute the filter that is passed to `.where`.

Copilot committed:
Fix dask-backed array issue in _sanitise_negative_uncertainties: compute filter before where

Agent-Logs-Url: https://github.com/openghg/openghg/sessions/18e8ecbc-c6e7-4897-abd1-b15918625988

Co-authored-by: brendan-m-murphy <11130776+brendan-m-murphy@users.noreply.github.com>

brendan-m-murphy committed:
Merge remote-tracking branch 'origin/devel' into Iss1538_negerror

brendan-m-murphy committed:
Address negative uncertainty review comments

SutarPrasad reviewed (approved):
looks good
